src/rust/src/file_reader.rs

```diff
@@ -28,13 +36,20 @@ pub struct FgbReader<R, State = Initial> {
     cur_pos: u64,
     /// All features read or end of file reached
     finished: bool,
-    /// Reader state
-    state: PhantomData<State>,
+    /// Whether or not the underlying reader is Seek
+    seekable_marker: PhantomData<S>,
+}
```

Comment:
Do you even need the `read_trait` parameter?

Isn't `R: Read + Seek` like in
```
impl<R: Read + Seek> FgbReader<R> {
    pub fn select_all(mut self) -> Result<FeatureIter<R>> {
```
already enough?

Reply:
I'd be happy to get rid of it, but I don't think we can.

Ultimately we want to have two different  `impl FallibleStreamingIterator` declarations for the FgbReader — one for `Read` and another one optimized for `Read + Seek`.

Because `Read + Seek` is a subtype of `Read`, we have to do something to make sure the implementations aren't overlapping, otherwise you'll see something like:

```
error[E0119]: conflicting implementations of trait `MyTrait` for type `MyStruct<_>`
   --> src/file_reader.rs:255:1
    |
249 | impl<R: Read> MyTrait for MyStruct<R> {
    | ------------------------------------- first implementation here
...
255 | impl<R: Read + Seek> MyTrait for MyStruct<R> {
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ conflicting implementation for `MyStruct<_>`
```

Because we want a distinct `impl FallibleStreamingIterator` for the thing returned by `select_all` vs `select_all_seq`, they must have distinct types.

We could introduce some new struct, e.g. `FeatureIterSeekable` and `FeatureIterNotSeekable`, but instead we return `FeatureIter<_, Seekable>`| `FeatureIter<_, NotSeekable>`.

I think this approach makes sharing common functionality a little easier. There's still a little duplicated code with this approach, but I think it's as minimal as it could be. I actually think it's kind of clever, and I don't think I would have come up with the idea if you hadn't introduced a similar concept in 30af03d0811607d7b40166e9bbd5efb69c830557, so thanks for the inspiration. =)